goodai/ltm/memory.py

```diff
@@ -0,0 +1,207 @@
+        return self._sync_call("retrieve", query=query, k=k, max_distance=max_distance)
+
+    def retrieve_from_keywords(self, keywords: list[str]) -> list[RetrievedMemory]:
+        assert len(keywords) > 0
+        return self._sync_call("retrieve_from_keywords", keywords=keywords)
+
+    def __end__(self):
+        self.mem_server.terminate()
+
+
+class LTMSystem:
+
+    def __init__(
+        self, chunk_capacity: int = 50, chunk_overlap_fraction=0, **other_params,
+    ):
+        self.semantic_memory = AutoTextMemory.create(config=TextMemoryConfig(
+            chunk_capacity=chunk_capacity,
+            chunk_overlap_fraction=chunk_overlap_fraction,
+            **other_params,
+        ))
+        self.keyword_index = defaultdict(list)
+
+    def is_empty(self) -> bool:
+        return self.semantic_memory.is_empty()
+
+    def clear(self):
+        self.semantic_memory.clear()
+
+    def state_as_text(self) -> str:
+        return json.dumps(dict(
+            semantic_memory=self.semantic_memory.state_as_text(),
+            keyword_index=self.keyword_index,
+        ))
+
+    def set_state(self, state_text: str):
+        state = json.loads(state_text)
+        self.keyword_index = defaultdict(list, state["keyword_index"])
+        self.semantic_memory.set_state(state["semantic_memory"])
+
+    def add_content(
+        self, content: str, timestamp: float = None, keywords: list[str] = None,
+        **metadata: Any,
+    ) -> int:
+        keywords = keywords or []
+        content, metadata = self.content_addition_preprocessing(content, keywords, **metadata)
+        text_key = self.semantic_memory.add_text(
+            content, timestamp=timestamp, metadata=metadata,
+        )
+        self.semantic_memory.add_separator()
+        for kw in keywords:
+            self.keyword_index[kw].append(text_key)
+        return text_key
+
+    def retrieve(
+        self, query: str, k: int, max_distance: float = None,
+    ) -> list[RetrievedMemory]:
+        memories = self.semantic_memory.retrieve(query, k)
+        if max_distance is not None:
+            memories = [m for m in memories if m.distance <= max_distance]
+        return memories
+
+    def retrieve_from_keywords(self, keywords: list[str]) -> list[RetrievedMemory]:
+        assert len(keywords) > 0
+        text_keys = set()
+        for kw in keywords:
+            text_keys.update(self.keyword_index[kw])
+        mem: DefaultTextMemory = self.semantic_memory
+        memories = list()
+        for chunk in mem.get_all_chunks():
+            assert len(chunk.associated_keys) == 1
+            text_key = chunk.associated_keys[0]
+            if text_key not in text_keys:
+                continue
+            text_keys.remove(text_key)
+            token_ids = mem.chunk_queue.get_sequence_token_ids(text_key)
+            text = mem.chunk_tokenizer.decode(token_ids, skip_special_tokens=True)
+            matching_keywords = set(chunk.metadata["keywords"]) & set(keywords)
```

Comment:
If the chunk has the metadata already associated with it, and you are iterating though all chunks, then why the `text_key`?

Reply:
One memory is associated with a `text_key` and every `text_key` with an undefined amount of chunks. However, there's no structure that holds a list of `text_key`s, but they're always part of the chunks, so I have to iterate through the chunks, and once I find the key I'm looking for I can take the metadata and add the key to the set, just for the sake of ignoring other chunks from the same memory.

Reply:
Okay, that makes sense for now. Only pitfall I can see is that if we stop doing the `chunk -> interaction` resolution for some reason, then the keyword retrieval will only get the first chunk of a memory that has that keyword. Its just something to keep in mind.

Reply:
Yeah, I think it's no problem as long as we add separators between memories. Each memory might be split in several chunks, but all of them should link to the same metadata.